### PR TITLE
Use MRI C ext headers

### PR DESCRIPTION
**Pull request by bjfish**
Steps are documented in `doc/contributor/updating-ruby.md` regarding how to update headers.

### CODE REVIEW

**Comment by chrisseaton** [comment]
Sorry, I need to pull this locally to review I think. May take me another day.

**Comment by bjfish** [comment]
One nice way to review is to do a folder compare (`truffleruby/lib/cext/ruby` with `ruby/include/ruby`) with the original MRI headers with a tool like kdiff3/beyond compare to see what we do/don't change.

**Comment by chrisseaton** [comment]
I do now think this is a good idea. But the list of regexes does not look maintainable! How about a patch?

**Comment by bjfish** [comment]
@chrisseaton I started development with regex replacement because I thought it might be that bulk style changes would be applied to the header or other bulk transformations.  I agree a tool like this could be confusing to maintain.

With a patch file, would we just replace these regexes with a patch apply? And if changes need to be made we edit/update the patch? Here is what the current patch file looks like:
https://gist.github.com/bjfish/c7825e1b8000a42f5513d68886e8ef1f

Maybe this regex tool could be used this one time to generate these initial headers and updates could be handled manually going forward?

**Comment by chrisseaton** [comment]
I think what we would do would be to apply that patch to the version in our repository and commit that.

Then we we upgrade the version of MRI we use we would copy in fresh copies of the headers, and apply the patch again, manually fixing any merge conflicts. This would effectively be a rebase - we would rebase our changes onto the new version of MRI.

I think the patch file is surely much more readable than the regexes.

**Comment by chrisseaton** [comment]
We wouldn't need to commit the patch itself - we could regenerate it from diffing against the version of MRI that we were previously using.

**Comment by bjfish** [comment]
@chrisseaton It sounds like this PR just needs to remove the tool then if it's not going to be used going forward.  I'll keep it for now to give others a chance to review.

**Comment by eregon** [comment]
One way is how JRuby does it for stdlib. They have a fork of MRI and then just keep commits above the chosen MRI version. When updating, it amounts to rebasing those commits on the new MRI version.

I think that's a nice approach, but it's tricky to keep that fork and the main repo in sync.
Doing everything in one repo might work as well, but then file renames can make getting the full diff from the original MRI headers very difficult, and the diff might grow quite large to a point it's a very hard to re-apply.

In any case, minimizing the diff is important.

**Comment by chrisseaton** [comment]
What's going on with this PR now? It's growing conflicts.

**Comment by bjfish** [comment]
@chrisseaton I think @eregon wanted to wait till you were back to look this over. I'll clean this up and get it ready for review again ...

**Comment by bjfish** [comment]
@chrisseaton @eregon I've re-organized this into three commits: first to add the tool and non-header changes, the second commit copies over the MRI 2.3.3 headers exactly, and then the 3rd commit applies our changes.

The 3rd commit is the most interesting: 6c4f57d4f2b1ec8957dfee16a9d3de73db2e876b
Being a commit, it is easy to make into a patch.  Would the following update process work then?
1. Create a patch from previous "patch" commit (e.g. 6c4f57d4f2b1ec8957dfee16a9d3de73db2e876b)
2. Replace all header files with the latest from MRI (2.4.x for example), (and commit them in original form?)
3. Apply the patch and resolve any conflicts, commit.

**Comment by chrisseaton** [comment]
I think I'd describe the process as:

1. Copy over the original version headers over ours.
2. Create a patch from this change.
3. Copy over the new version headers over ours.
4. Apply the patch.

**Comment by eregon** [comment]
But then 2. is a reversed patch from ours to vanilla.
I think committing the original headers is much cleaner, then the diff is in the natural direction.

File: lib/cext/ruby/ruby.h [hunk 1]
```diff
@@ -2215,6 +2118,319 @@ void ruby_sig_finalize(void);
...
+
+  if (found_kwargs) {
+    return argc - 1;
+  } else {
+    return argc;
+  }
+}
+
+
+
```
Comment (eregon) [inline lib/cext/ruby/ruby.h]: I would split the huge number of additions we have here in another file like `lib/cext/truffle/ruby.h`, so then the diff is just one `#include` line.

File: lib/cext/ruby/ruby.h [hunk 2]
```diff
@@ -1833,185 +1747,189 @@ VALUE rb_yield_splat(VALUE);
...
-RUBY_EXTERN VALUE rb_eEncCompatError;
-
-RUBY_EXTERN VALUE rb_eScriptError;
-RUBY_EXTERN VALUE rb_eNameError;
-RUBY_EXTERN VALUE rb_eSyntaxError;
-RUBY_EXTERN VALUE rb_eLoadError;
-
-RUBY_EXTERN VALUE rb_eMathDomainError;
-
-RUBY_EXTERN VALUE rb_stdin, rb_stdout, rb_stderr;
```
Comment (eregon) [inline lib/cext/ruby/ruby.h]: We could actually use these global variables, and initialize them with a function we define.
We could use `__attribute__((constructor))` or just our own function which we call when loading ruby.su.

File: test/mri/excludes/TestNum2int.rb [hunk 3]
```diff
@@ -7,3 +7,5 @@
 exclude :test_num2uint, "needs investigation"
 exclude :test_num2ulong, "needs investigation"
 exclude :test_num2ushort, "needs investigation"
+exclude :test_num2ll, "needs investigation"
+exclude :test_num2ull, "needs investigation"
```
Comment (eregon) [inline test/mri/excludes/TestNum2int.rb]: Could you split these changes in another commit, or maybe just reuse this one but rename and remove the tool?

File: src/main/c/cext/ruby.c [hunk 4]
```diff
@@ -1595,7 +1602,7 @@ int rb_is_instance_id(ID id) {
 
 // Array
 
-long rb_array_len(VALUE array) {
+int RARRAY_LEN(VALUE array) {
```
Comment (eregon) [inline src/main/c/cext/ruby.c]: I think this is a rebase error.

**Comment by eregon** [comment]
Having the diff in the log of git is quite useful and I think has much better context than a raw patch file, which becomes hard to apply with just a few changes or a bit of time.

**Comment by bjfish** [comment]
@eregon I don't think the tool is needed going forward, I've removed it from the first commit and updated the message.

File: lib/cext/ruby/encoding.h [hunk 5]
```diff
@@ -349,63 +348,65 @@ VALUE rb_econv_append(rb_econv_t *ec, const char *bytesrc, long bytesize, VALUE
...
+      RUBY_ECONV_STATEFUL_DECORATOR_MASK          = 0x00f00000,
+      RUBY_ECONV_XML_ATTR_QUOTE_DECORATOR         = 0x00100000,
+  
+      RUBY_ECONV_DEFAULT_NEWLINE_DECORATOR        = 0x00000000,
+      /* end of flags for rb_econv_open */
+  
+      /* flags for rb_econv_convert */
+      RUBY_ECONV_PARTIAL_INPUT                    = 0x00010000,
+      RUBY_ECONV_AFTER_OUTPUT                     = 0x00020000,
+      /* end of flags for rb_econv_convert */
```
[the remark this answers is not in the document]
Comment (bjfish) [reply lib/cext/ruby/encoding.h]: Fixed.

File: lib/cext/ruby/ruby.h (hunk 2 above)
Comment (bjfish) [reply lib/cext/ruby/ruby.h]: This sounds like a nice idea. I think it's outside of the scope of this PR though.

File: test/mri/excludes/TestNum2int.rb (hunk 3 above)
Comment (bjfish) [reply test/mri/excludes/TestNum2int.rb]: Done. Tool removed and message updated.

File: src/main/c/cext/ruby.c (hunk 4 above)
Comment (bjfish) [reply src/main/c/cext/ruby.c]: Yes, rebase error, fixed.

**Comment by chrisseaton** [comment]
However you do the diff and update, please update `doc/contributor/updating-ruby.md` to describe it.

**Comment by bjfish** [comment]
@chrisseaton I've documented the steps to update in `doc/contributor/updating-ruby.md` and tested these steps locally using a later version of the headers. The steps seem pretty straightforward and manageable.

File: lib/cext/ruby/ruby.h (hunk 2 above)
Comment (eregon) [reply lib/cext/ruby/ruby.h]: Yes, absolutely.

**Comment by eregon** [comment]
Looks good to me, thanks and sorry it took so long to integrate.

Could you gist the tool and post the URL here?
It might be useful one day, even if we won't update it.

**Comment by bjfish** [comment]
@eregon Gist of the tool here: https://gist.github.com/bjfish/534f99282ed785a61752a566a244a359
